### ISSUE TITLE
Define the high level features of the BC Registries Event Processor

### ISSUE THREAD

**Issue by swcurran** [bcgov/von-bc-registries-agent#8]

**Comment by jljordan42** [comment]
@ianco initial design thoughts 

- BC Registries/Event Processor https://github.com/ianco/von-bc-registries-agent/tree/master/doc

**Comment by swcurran** [comment]
It's tough to understand entirely what you mean in the design, so we should walk through it.  One thing that we can't do is simply process events of a certain type, as this seems to be suggesting.  For the first event about an organization pushed TOB we must have the very latest information about the organization - the data for the current state of the organization.  As such, perhaps a better approach for initial loading is to find the last event from each organization and process that so that TOB holds the current "active" credential.  Alternatively - we simply start at the latest issued Credential and work forward and backward through the database. That is - with each call we get any new events (forward), ignore the ones we have and get some number (100, 1000, whatever) of the events before our last one (backward). 

Of course going back assumes we can get the content of the Credentials as they were at the time of the issuance.  If we can't get that, it's pretty well pointless to process anything except the most recent event.  From the design document, it's not clear we can get "point in time" information about an Entity.

Anyway - things to discuss.  I think we have to sit down face to face to discuss.

**Comment by swcurran** [comment]
Regards the handling of the events, data fetch and credentials creation.  My preference would be that the event gets picked up and logged on a queue with a minimum of information - e.g. don't collect the point in time details about the event at that time.  At the time the event gets processed by a worker handling the queue, get the data for it, map it to the credential and issue the credential.  Once done, mark the event complete on the queue.

My rationale is that this pattern matches a webhook style event processing - something happens, minimal processing is done to get the event queued for processing, and then when it is processed, the heavier weight effort around what needs to be done can happen.  Likely, the event processing can be handled by a single threaded process, while the workers can be scaled up as needed to do all the heavy lifting for a specific event.

That also allows the Worker process to dig into what should happen - applying some business rules on ignoring some events, getting different data for other events, etc.  I'm thinking about, for example, the handling of the initial loading events (for each company, find the latest data associated with that company's latest event - not necessarily the one found in the queue), and when we are going back in history, figuring out how to set the "enddate" on a credential that we are using for completeness but is already obsolete based on more recent events.

In other words - fetching data while adding the event may be wasted effort, and should be left to a Worker process.

Not earth-shattering, but that's what feels right to me.

**Comment by ianco** [comment]
Makes sense.  I'm packaging the BC Registries access functions in a stand-alone class, so it will be easy to patch into whatever step of the process we feel is appropriate.  I'll implement as per your notes.

**Comment by ianco** [comment]
https://github.com/mara/data-integration
